test/extensions/filters/network/thrift_proxy/conn_manager_test.cc

```diff
@@ -1058,6 +1085,178 @@ TEST_F(ThriftConnectionManagerTest, ResetDownstreamConnection) {
+  EXPECT_EQ(0U, store_.counter("test.response_error").value());
+}
+
+// Test the case where there is a limit on the number of requests and the actual number of requests
+// exceeds the limit.
+TEST_F(ThriftConnectionManagerTest, RequestWithMaxRequestsLimitAndReachedWithMoreRequests) {
+  const std::string yaml = R"EOF(
+    stat_prefix: test
+    route_config:
+      name: local_route
+    max_requests_per_connection: 50
+    )EOF";
+
+  initializeFilter(yaml);
+  EXPECT_EQ(50, config_->maxRequestsPerConnection());
+
+  // Since max requests per connection is set to 50, the connection will be disconnected after
+  // all 50 requests is completed.
+  EXPECT_CALL(filter_callbacks_.connection_, close(Network::ConnectionCloseType::FlushWrite));
+
+  EXPECT_EQ(50, sendSomeThriftRequest(55));
+
+  filter_callbacks_.connection_.dispatcher_.clearDeferredDeleteList();
+
+  EXPECT_EQ(1U, store_.counter("test.downstream_cx_max_requests").value());
+  EXPECT_EQ(50U, store_.counter("test.request").value());
+  EXPECT_EQ(50U, store_.counter("test.request_call").value());
+  EXPECT_EQ(0U, stats_.request_active_.value());
+  EXPECT_EQ(50U, store_.counter("test.response").value());
+  EXPECT_EQ(50U, store_.counter("test.response_reply").value());
+  EXPECT_EQ(0U, store_.counter("test.response_exception").value());
+  EXPECT_EQ(0U, store_.counter("test.response_invalid_type").value());
+  EXPECT_EQ(50U, store_.counter("test.response_success").value());
+  EXPECT_EQ(0U, store_.counter("test.response_error").value());
+}
+
+// Test cases where the number of requests is limited and the actual number of requests exceeds the
+// limit several times.
+TEST_F(ThriftConnectionManagerTest, RequestWithMaxRequestsLimitAndReachedRepeatedly) {
+  const std::string yaml = R"EOF(
+    stat_prefix: test
+    route_config:
+      name: local_route
+    max_requests_per_connection: 5
+    )EOF";
+
+  initializeFilter(yaml);
+  EXPECT_EQ(5, config_->maxRequestsPerConnection());
+
+  EXPECT_CALL(filter_callbacks_.connection_, close(Network::ConnectionCloseType::FlushWrite))
+      .Times(5);
+
+  auto mock_new_connection = [this]() {
+    filter_callbacks_.connection_.dispatcher_.clearDeferredDeleteList();
+
+    filter_ = nullptr;
+
+    filter_callbacks_.connection_.read_enabled_ = true;
+    filter_callbacks_.connection_.state_ = Network::Connection::State::Open;
+    filter_callbacks_.connection_.callbacks_.clear();
+
+    ON_CALL(random_, random()).WillByDefault(Return(42));
+    filter_ = std::make_unique<ConnectionManager>(
+        *config_, random_, filter_callbacks_.connection_.dispatcher_.timeSource());
+    filter_->initializeReadFilterCallbacks(filter_callbacks_);
+    filter_->onNewConnection();
+
+    filter_->onAboveWriteBufferHighWatermark();
+    filter_->onBelowWriteBufferLowWatermark();
+  };
+
+  EXPECT_EQ(5, sendSomeThriftRequest(6));
+
+  for (size_t i = 0; i < 4; i++) {
+    mock_new_connection();
+    EXPECT_EQ(5, sendSomeThriftRequest(6));
+  }
```

Comment:
Why is this loop needed?

Reply:
In order to simulate the situation of multiple disconnections, observe whether the stat value grows correctly.